### PR TITLE
3905 feature flag setup

### PR DESCRIPTION
**Pull request by johanna-skylight**
# BACKEND PULL REQUEST

## Related Issue

Implements feature flags that can be controlled from the DB table feature_flag ( #3905 )

## Changes Proposed
Diagram below showcases the setup and the elements that got created as part of this PR.

SR Feature Flags (1)

After this PR gets approved and merged to main I will include this diagram in the wiki.

## Testing

- Verify front end calls REST endpoint /feature-flags
- Verify endpoint returns multiplexEnabled flag set to true (and after changing the value in the table the response reflects those values after the 1 min refresh)
- Verify the frontend displays or hides multiplex features accordingly.

## Checklist for Primary Reviewer

- [X] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke tested
- [X] Any content updates (user-facing error messages, etc) have been approved by content team
- [X] Any changes that might generate questions in the support inbox have been flagged to the support team
- [X] GraphQL schema changes are backward compatible with older version of the front-end
- [X] Changes comply with the SimpleReport Style Guide
- [X] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [X] Any dependencies introduced have been vetted and discussed
- [X] Any changes to the startup configuration have been documented in the README

### CODE REVIEW

**Review by fzhao99 — DISMISSED** [review]
Working locally for me and looks really great! Thanks for the work on this Johanna!

**Review by emyl3 — COMMENTED** [review]
Do we want to clean up the use of `process.env.REACT_APP_MULTIPLEX_ENABLED` in `QueueItem.test.tsx`, `TestResultDetailsModal.test.tsx`, and `TestResult.test.tsx` as well in this PR? 🧐 

If so, do we also want to remove the `REACT_APP_MULTIPLEX_ENABLED` value from our `frontend/.env` file as well? 🤔 

Thank you for all your work on this!! 😸

**Comment by emyl3** [comment]
Oh right... Forgot to say THANK YOU for the diagram! It was very helpful 🤩 🌈

File: backend/src/main/java/gov/cdc/usds/simplereport/config/OktaLocalSecurityConfiguration.java [hunk 1]
```diff
@@ -66,6 +66,10 @@ protected void configure(HttpSecurity http) throws Exception {
         .antMatchers(HttpMethod.POST, WebConfiguration.TWILIO_CALLBACK)
         .permitAll()
 
+        // Feature Flags that apply at app level
```
[the remark this answers is not in the document]
Comment (johanna-skylight) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/OktaLocalSecurityConfiguration.java:69]: It seems (based on the behavior of my e2e testing issue) that it is being used only for the e2e testing, for the deployments in prod and test environments we are using the SecurityConfiguration.java file. So when adding a new REST endpoint both files need to be updated with the same change... (I think)

**Comment by johanna-skylight** [comment]
> Do we want to clean up the use of `process.env.REACT_APP_MULTIPLEX_ENABLED` in `QueueItem.test.tsx`, `TestResultDetailsModal.test.tsx`, and `TestResult.test.tsx` as well in this PR? 🧐
> 
> If so, do we also want to remove the `REACT_APP_MULTIPLEX_ENABLED` value from our `frontend/.env` file as well? 🤔
> 
> Thank you for all your work on this!! 😸

I think I should... let me clean those references. @emmastephenson can I remove the flag from the .env? I just want to double check with you.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/OktaLocalSecurityConfiguration.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/OktaLocalSecurityConfiguration.java:69]: I believe it's also used when running the local app with Okta enabled!

File: backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+class FeatureFlagTest {
+  FeatureFlag featureFlag = new FeatureFlag();
+
+  @Test
+  void equals_sameObject_returnsTrue() {
+    assertEquals(featureFlag, featureFlag);
+  }
```
Comment (BobanL) [inline backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java:15]: Is this just testing the lombok annotations? If yes, is this done for sonarcloud?

Comment (johanna-skylight) [reply backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java:15]: I just added that one because I saw it in other test classes for other Entity/Model Objects 🤷‍♀️ . I don't have a good answer... I was just trying to follow the pattern.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+
+@Component
+@ConfigurationProperties(prefix = "features")
+@EnableScheduling
+@Data
```
Comment (BobanL) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java:18]: Does this need  everything from`@Data`? Could this be replaced with just the annotations we need like `@Setter`?

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/FeatureFlag.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+public class FeatureFlag extends AuditedEntity {
+  @Column(nullable = false)
+  private String name;
+
+  @Column(nullable = false)
+  private Boolean value;
+
+  public FeatureFlag() {
+    /* required for hibernate*/
+  }
```
Comment (BobanL) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/FeatureFlag.java:20]: Can we remove this and just use the `@NoArgsConstructor`?

File: backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java (hunk 2 above)
Comment (BobanL) [reply backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java:15]: I think as long as it doesn't impact sonarcloud (not sure how to check actually) it should be ok to remove I think. I don't see a ton of value in testing Lombok annotations.  

If this is a pattern the team does follow regularly, then I'm ok with this.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java (hunk 3 above)
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java:18]: +1, I don't think we need to override the equals and hash constructor here. The `@setter` and potentially `@requiredArgsConstructor` should be enough.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+public class FeatureFlagsConfig {
+  @Autowired
+  @Getter(AccessLevel.NONE)
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java:21]: Removing `@data` and switching to specific getters and setters would also eliminate the need to use the access levels here I believe.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    Iterable<FeatureFlag> flags = _repo.findAll();
+    flags.forEach(
+        flag -> {
+          flagMapping(flag.getName(), flag.getValue());
+        });
+  }
+
+  private void flagMapping(String flagName, Boolean flagValue) {
+    switch (flagName) {
+      case "multiplexEnabled":
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java:38]: Does this mean that for future feature flags, we'll need to update the database with the flag and edit this mapping as well?

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/FeatureFlag.java (hunk 4 above)
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/FeatureFlag.java:20]: +1, we probably also want to add an `@RequiredArgsConstructor` so that it's possible to add new feature flags outside of a manual database write.

File: backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java (hunk 2 above)
Comment (emmastephenson) [reply backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java:15]: @johanna-skylight just for your SA, I think this is only done in other classes to test that overriding the `.equals` of a given object is working as intended. So we don't need it here because you're not changing the default equality.

File: backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
+package gov.cdc.usds.simplereport.db.model;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+class FeatureFlagTest {
+  FeatureFlag featureFlag = new FeatureFlag();
```
Comment (emmastephenson) [inline backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java:10]: Can you change this variable to be `private`? Otherwise it will default to `protected`, which _shouldn't_ be an issue but better to be cautious

File: frontend/src/app/supportAdmin/PendingOrganizations/PendingOrganizationsContainer.test.tsx [hunk 8]
```diff
@@ -488,7 +488,12 @@ describe("PendingOrganizationsContainer", () => {
           await waitForElementToBeRemoved(
             screen.queryByText("Organization details")
           );
-          expect(screen.getByText("DC Space Camp")).toBeInTheDocument();
+
+          expect(
+            screen.getByRole("heading", {
+              name: /dc space camp details updated/i,
```
Comment (emmastephenson) [inline frontend/src/app/supportAdmin/PendingOrganizations/PendingOrganizationsContainer.test.tsx:494]: Is this change related to this PR? If not could you make a separate change for it?

File: frontend/src/featureFlags/WithFeatureFlags.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import { FlagsProvider } from "flagged";
+
+import FetchClient from "../app/utils/api";
+
+type WithFeatureFlagsProps = {
+  children: JSX.Element;
+};
+
+const SR_APP_FEATURES = "sr-app-features";
+const api = new FetchClient(undefined, { mode: "cors" });
```
Comment (emmastephenson) [inline frontend/src/featureFlags/WithFeatureFlags.tsx:11]: Our pattern elsewhere has been to specify the root path at the `FetchClient` level (see [`AccountCreationApiService`](https://github.com/CDCgov/prime-simplereport/blob/c3e6b2dc330021e425e01657eaee613c51f9e11c/frontend/src/app/accountCreation/AccountCreationApiService.ts#L5-L5), could you edit this `FetchClient` to operate the same way?

File: frontend/src/index.tsx [hunk 10]
```diff
@@ -119,26 +120,28 @@ export const ReactApp = (
...
-                <Route path="/reload-app" element={<Navigate to="/" />} />
-                <Route path="/*" element={<App />} />
-                <Route element={<>Page not found</>} />
-              </Routes>
-            </PrimeErrorBoundary>
-          </TelemetryProvider>
-        </Router>
+        <WithFeatureFlags>
+          <Router basename={process.env.PUBLIC_URL}>
+            <TelemetryProvider>
```
Comment (emmastephenson) [inline frontend/src/index.tsx:125]: Before this merges, we should double-check that it doesn't interfere with Azure's Telementry reporting at all...it shouldn't, but I don't really trust Azure to do the sensible thing anymore

**Review by emmastephenson — COMMENTED** [review]
Thanks for all your work on this Johanna!! Please let me know if you have questions about any of the comments I left.

**Comment by emmastephenson** [comment]
Also @johanna-skylight, once you've made the next round of changes, can you add @rin-skylight as a reviewer for the security side? I don't see any potential security flags atm but I am not the security expert on the team 😄

File: frontend/src/app/supportAdmin/PendingOrganizations/PendingOrganizationsContainer.test.tsx (hunk 8 above)
Comment (johanna-skylight) [reply frontend/src/app/supportAdmin/PendingOrganizations/PendingOrganizationsContainer.test.tsx:494]: I added this change because the frontend unit test randomly started failing with an error of not being able to find the text "DC Space Camp" so in order to pass the unit test I had to change it.

File: frontend/src/index.tsx (hunk 10 above)
Comment (johanna-skylight) [reply frontend/src/index.tsx:125]: How can that be tested 🤔 ?

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/FeatureFlag.java (hunk 4 above)
Comment (johanna-skylight) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/FeatureFlag.java]: The @NoArgsConstructor was good but when I tried adding @RequiredArgsConstructor the project started to throw exceptions during build. Can this be added in the feature as need arises?

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java (hunk 6 above)
Comment (johanna-skylight) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java:38]: yes, as the flagConfig object and the featureFlag object are not 1 to 1. I think this mapping will be handing when expanding support of feature flags  by facility and org

File: frontend/src/index.tsx (hunk 10 above)
Comment (emmastephenson) [reply frontend/src/index.tsx:125]: We'll need to deploy to a testing environment and then check that metrics are still coming in through for frontend events. I'm happy to pair with you on doing that!

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+
+@Component
+@ConfigurationProperties(prefix = "features")
+@EnableScheduling
+@Setter
+@Getter
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java:18]: I believe that instead of setting these at the class level, you could just set them for the `multiplexEnabled` field. If I'm understanding the `AccessLevel.NONE` enum correctly, that would accomplish the same thing - no getters/setters on the _repo, but you'd still have them for the `multiplexEnabled` bool.

File: backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
+package gov.cdc.usds.simplereport.db.model;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+public class FeatureFlagTest {
```
Comment (emmastephenson) [inline backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java:9]: Can you remove the public modifier here?

File: backend/src/test/java/gov/cdc/usds/simplereport/api/featureflags/FeatureFlagsControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  @BeforeEach
+  public void setup() {
+    _mockFeatureFlagConfig.setMultiplexEnabled(true);
+    this.featureFlagsController = new FeatureFlagsController();
+    ReflectionTestUtils.setField(
+        this.featureFlagsController, "featureFlags", _mockFeatureFlagConfig);
+  }
+
+  @Test
+  public void endpointReturnsFeatureFlagsConfigObj() {
```
Comment (emmastephenson) [inline backend/src/test/java/gov/cdc/usds/simplereport/api/featureflags/FeatureFlagsControllerTest.java:24]: Can you remove the `public` modifier on this test?

Comment (johanna-skylight) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/featureflags/FeatureFlagsControllerTest.java:24]: Removing the public modifier makes random backend test start failing so I am not sure how how to move forward without it

File: backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java (hunk 12 above)
Comment (johanna-skylight) [reply backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java:9]: Similar to my comment above. If I set the feature flag field as private and I do not add the public modifier to the class some random not related unit test in the backend start failing.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java (hunk 11 above)
Comment (johanna-skylight) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java:18]: My logic behind adding it at the class level is that when the config object grows we will not need to add the getter and the setter annotation at the field level. I am predicting that we will need to be able to access the individual flags /fields across the app (when using flags in the Java code) and the repo object is the only one that should stay private in the class. 

So that is why I went with the approach of add all fields minus _repo.

File: frontend/src/featureFlags/WithFeatureFlags.tsx (hunk 9 above)
Comment (johanna-skylight) [reply frontend/src/featureFlags/WithFeatureFlags.tsx]: Done!

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java (hunk 11 above)
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/FeatureFlagsConfig.java:18]: I see, that makes sense for now. Thank you!

File: backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java (hunk 12 above)
Comment (emmastephenson) [reply backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java:9]: I dug into this a little more, and at this point I think it's a framework issue of some kind - we are using this `graphql-java-kickstart:graphql-spring-boot-starter` library to integrate graphQL into our setup, and this failure seems to be something of a known issue there: https://github.com/graphql-java-kickstart/graphql-spring-boot/discussions/419

I haven't been able to link it specifically to the use of `public`, but now that there's an official [SpringBoot/GraphQL integration](https://spring.io/projects/spring-graphql) I think we'd be better served by just migrating to that instead of trying to continue debugging here. Our test setup (`BaseGraphQLTest`) is also unlike any other test setup I've seen on the web...so that will be a fun puzzle to unravel once we start doing the migration.

So @johanna-skylight, if you could leave a comment on these files saying we should revisit `public` after the Spring for GraphQL migration is done, then I think we'll call this good for now. Thanks for your patience in investigating!

**Review by rin-skylight — CHANGES_REQUESTED** [review]
I have found a potential issue that needs to be addressed.

In the browser, the use of local storage is not secure. A user can easily modify the existing value of any feature flag.

On our `test` environment, I attempted to manipulate the value of the feature flag by manually setting it to "false". It does not seem like that has any effect on the app, as switching screens does not cause any change to the UI elements displayed. Upon a refresh, the app will hit the API endpoint to update the flag value, so I wonder whether the use of local storage is fully necessary.

In a bit of a more nefarious attempt, however, I replaced the value of the feature flag with random characters, and attempted to refresh the app. This time, the feature flag was not updated with the latest data from the API endpoint, and multiplex features did not display. I was able to conduct a test with a multiplex device, save just the COVID-19 results, and then recall the data with missing results for Flu A and Flu B.

It seems like there's a bit of a gap here...a user should not be able to change the functionality of the application, or the integrity of the data, through client-side manipulations. I think we need either stronger server-side validation, or we should reconsider making this setting available via local storage.

File: frontend/src/index.tsx (hunk 10 above)
Comment (rin-skylight) [reply frontend/src/index.tsx:125]: Can someone confirm once this is complete? I'd really appreciate it!

File: backend/src/test/java/gov/cdc/usds/simplereport/api/featureflags/FeatureFlagsControllerTest.java (hunk 13 above)
Comment (johanna-skylight) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/featureflags/FeatureFlagsControllerTest.java:25]: added comments

File: backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java (hunk 12 above)
Comment (johanna-skylight) [reply backend/src/test/java/gov/cdc/usds/simplereport/db/model/FeatureFlagTest.java:10]: added comments

File: frontend/src/app/supportAdmin/PendingOrganizations/PendingOrganizationsContainer.test.tsx (hunk 8 above)
Comment (johanna-skylight) [reply frontend/src/app/supportAdmin/PendingOrganizations/PendingOrganizationsContainer.test.tsx]: This test was only failing in the CI (local always passed the test). Seems that root cause was that jest was not properly waiting for the DOM to be updated and failing the test too soon. I have restored the former assertion but added waiting code to do the assertion async.

**Comment by johanna-skylight** [comment]
> I have found a potential issue that needs to be addressed.
> 
> In the browser, the use of local storage is not secure. A user can easily modify the existing value of any feature flag.
> 
> On our `test` environment, I attempted to manipulate the value of the feature flag by manually setting it to "false". It does not seem like that has any effect on the app, as switching screens does not cause any change to the UI elements displayed. Upon a refresh, the app will hit the API endpoint to update the flag value, so I wonder whether the use of local storage is fully necessary.
> 
> In a bit of a more nefarious attempt, however, I replaced the value of the feature flag with random characters, and attempted to refresh the app. This time, the feature flag was not updated with the latest data from the API endpoint, and multiplex features did not display. I was able to conduct a test with a multiplex device, save just the COVID-19 results, and then recall the data with missing results for Flu A and Flu B.
> 
> It seems like there's a bit of a gap here...a user should not be able to change the functionality of the application, or the integrity of the data, through client-side manipulations. I think we need either stronger server-side validation, or we should reconsider making this setting available via local storage.

The local storage caching was put in place to address the edge case of a user using the application with multiple tabs while dealing with slow internet connections. After syncing up with @emmastephenson it was decided that this edge case was not a priority so I have removed all local storage related functionality. I have also updated the diagram!

File: backend/src/test/java/gov/cdc/usds/simplereport/api/featureflags/FeatureFlagsControllerTest.java [hunk 14]
```diff
@@ -7,6 +7,7 @@
 import org.junit.jupiter.api.Test;
 import org.springframework.test.util.ReflectionTestUtils;
 
+/** ToDo Remove public modifiers from this test class once Spring for GraphQL migration is done */
```
Comment (emmastephenson) [inline backend/src/test/java/gov/cdc/usds/simplereport/api/featureflags/FeatureFlagsControllerTest.java:10]: Thanks for adding these comments!

**Review by emmastephenson — APPROVED** [review]
LGTM, thanks for your work on this Johanna! 

Could you put the diagram + a short explanation of the feature flag flow in our wiki when you have time?

File: frontend/src/index.tsx (hunk 10 above)
Comment (emyl3) [reply frontend/src/index.tsx:125]: @johanna-skylight As I'm reviewing your PR, wanted to follow up on this ask. Please let me know if I can help check since I have Azure access now 😸

Comment (emmastephenson) [reply frontend/src/index.tsx:125]: Just followed up on this and the telemetry appears unaffected - events are still being logged in Azure. I think this PR is good to go from a metrics perspective.

Comment (emyl3) [reply frontend/src/index.tsx:125]: Thank you for checking! That was the only thing for me and now the merge conflict I introduced when I merged in my PR a couple days ago 😅  -- otherwise this looks ready to go to me too @johanna-skylight

**Comment by johanna-skylight** [comment]
@emmastephenson @emyl3 I resolved the merged conflict. I hope this PR is good to go now!

**Review by emyl3 — APPROVED** [review]
Thank you for all your hard work on this, Johanna!! 😭 🎉

**Comment by johanna-skylight** [comment]
@rin-skylight can you review the PR when you get a chance? The changes for LocalStorage have been pushed so on that front we are good.

**Review by rin-skylight — APPROVED** [review]
Thank you for addressing all the feedback! This looks great!
